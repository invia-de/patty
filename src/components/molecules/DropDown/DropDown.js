import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

import cx from '../../../utils/classnames';

import styles from './dropdown.module.scss';

/**
 * @author [Roman Semko](mailto:roman.semko-extern@invia.de)
 * @since 0.1.0
 */
class DropDown extends React.Component {
  constructor() {
    super();
    this.ref = React.createRef();
    this.state = {
      isOpen: false,
      keep: false
    };
    this.onOpen = this.onOpen.bind(this);
    this.onClose = this.onClose.bind(this);
    this.onClick = this.onClick.bind(this);
    this.onClickOutside = this.onClickOutside.bind(this);
  }

  componentDidMount() {
    document.addEventListener('click', this.onClickOutside, true);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.onClickOutside, true);
  }

  render() {
    const {
      className,
      classNameHandler,
      handler,
      children,
      openBackgroundColor,
      align
    } = this.props;
    const { isOpen } = this.state;
    const cls = defaultClassName =>
      cx(defaultClassName, isOpen ? styles.open : '');
    const openStyle =
      isOpen && openBackgroundColor && openBackgroundColor.length > 0
        ? { backgroundColor: openBackgroundColor }
        : {};

    return (
      <div
        className={styles.dropdown}
        onMouseEnter={this.onOpen}
        onMouseLeave={this.onClose}
        ref={ref => {
          this.ref = ref;
        }}
      >
        <div
          className={cls(cx(styles.handler, classNameHandler))}
          onClick={this.onClick}
          style={openStyle}
        >
          {React.cloneElement(handler, { isOpen })}
        </div>
        {isOpen && (
          <div
            className={cls(
              cx(
                styles.content,
                align === 'right' ? styles.right : '',
                className
              )
            )}
            style={openStyle}
          >
            {children}
          </div>
        )}
      </div>
    );
  }

  onOpen() {
    const { openOnHover } = this.props;

    openOnHover && this.setState({ isOpen: true });
  }

  onClose() {
    const { openOnHover, keepOnClick } = this.props;
    const { keep } = this.state;

    openOnHover && (!keepOnClick || !keep) && this.setState({ isOpen: false });
  }

  close() {
    this.setState({ isOpen: false, keep: false });
  }

  onClick() {
    const { keepOnClick } = this.props;
    const { isOpen, keep } = this.state;

    keepOnClick &&
      this.setState({
        keep: !keep,
        isOpen: keep || (keepOnClick && !isOpen) ? !isOpen : isOpen
      });
  }

  onClickOutside(e) {
    if (this.ref && !this.ref.contains(e.target)) {
      this.setState({ isOpen: false, keep: false });
    }
  }
}
DropDown.propTypes = {
  /** additional classNames to be applied to the handler and dropdown content containers when the dropdown is open */
  className: PropTypes.string,
  classNameHandler: PropTypes.string,
  /** required due to accessibility */
  children: PropTypes.node.isRequired,

  /** react element to be used as handler */
  handler: PropTypes.any.isRequired,
  /** whether to open/close the dropdown on mouse hover */
  openOnHover: PropTypes.bool,
  /** control dropdown open/close with clicks on the handler */
  keepOnClick: PropTypes.bool,
  /** dropdown content alignment relative to the handler */
  align: PropTypes.oneOf(['left', 'right']),
  /** background color to be applied to the handler and content containers when the dropdown is open */
  openBackgroundColor: PropTypes.string
};

DropDown.defaultProps = {
  align: 'left',
  openOnHover: true,
  openBackgroundColor: ''
};

export default DropDown;

export function renderDropDown(props, container, callback) {
  ReactDOM.render(<DropDown {...props} />, container, callback);
}
