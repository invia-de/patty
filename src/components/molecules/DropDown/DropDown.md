A dropdown consists of a **handler** that opens/closes the dropdown on hover or click.

Simple dropdown that opens/closes on hover:

```jsx
const handler = <div style={{ padding: 6 }}>Hover over me to open/close!</div>;

<DropDown handler={handler} openBackgroundColor="#e2001a">
  <ul style={{ width: 250, color: 'white' }}>
    <li>First element</li>
    <li>Second element</li>
    <li>Third element</li>
    <li>Forth element</li>
  </ul>
</DropDown>;
```

It is possible to define a `openBackgroundColor` which will be applied to the
`handler`and dthe dropdown `content` containers, when the dropdown is open.

The style of handler and content are controlled from outside, as both are passed
as react elements to the `DropDown` container.

With `keepOnClick` it is possible to define that the dropdown should be kept open, even when moving out
with the mouse, if the handler is clicked. To close the dropdown the handler should be clicked again or
click anywhere outside of the dropdown.

It is also possible to align the dropdown to the right border of the handler.

Both examples:

```jsx
const handler = (
  <div style={{ padding: 6 }}>Toggle on hover, keep with a click!</div>
);

<DropDown handler={handler} keepOnClick align="right">
  <ul style={{ width: 250, color: 'white' }}>
    <li>First element</li>
    <li>Second element</li>
    <li>Third element</li>
    <li>Forth element</li>
  </ul>
</DropDown>;
```

It is also possible to open/close the dropdown with clicks, disabling the hover
completely with `openOnHover={false}`:

```jsx
const handler = <div style={{ padding: 6 }}>Click me to open/close!</div>;

<DropDown
  handler={handler}
  keepOnClick
  openOnHover={false}
  openBackgroundColor="#15cfff"
>
  <ul style={{ width: 250, color: 'white' }}>
    <li>First element</li>
    <li>Second element</li>
    <li>Third element</li>
    <li>Forth element</li>
  </ul>
</DropDown>;
```
