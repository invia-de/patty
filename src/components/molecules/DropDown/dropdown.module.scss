@import '../../../style/colors';
@import '../../../style/media-queries';

.dropdown {
  display: inline-block;
  position: relative;
  cursor: pointer;
  .handler {
    user-select: none;
  }
  .content {
    position: absolute;
    padding: 12px;
    @include respond-below(sm) {
      position: static;
      width: 100%;
      min-height: 100%;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
    }
  }

  .right {
    right: 0;
  }

  .open {
    background-color: $blue-dark;
  }
}
