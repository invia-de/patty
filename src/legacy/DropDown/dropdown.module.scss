@import '../variables.scss';

.dropdown {
  display: inline-block;
  position: relative;
  cursor: pointer;
  .handler {
    user-select: none;
  }
  .content {
    position: absolute;
    padding: 12px;
  }

  .right {
    right: 0;
  }

  .open {
    background-color: $blue-200;
  }
}
