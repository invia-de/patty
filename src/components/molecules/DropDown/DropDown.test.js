import React from 'react';
import DropDown from './DropDown';
import { render } from 'react-testing-library';

test('DropDown renders', () => {
  const { container } = render(
    <DropDown handler={<span>hi</span>}>
      <div>pass test for now</div>
    </DropDown>
  );
  expect(container.firstChild).toMatchSnapshot();
});
