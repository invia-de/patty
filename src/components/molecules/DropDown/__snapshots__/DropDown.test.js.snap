// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown renders 1`] = `
<div
  class="dropdown"
>
  <div
    class="handler "
  >
    <span>
      hi
    </span>
  </div>
</div>
`;
